### PR TITLE
Fixed #11 -- Switched to use irc3 lib

### LINKED ISSUE #11: Try using IRC SASL auth to circumvent Freenode Blacklist of AWS IP addresses
It seems this is the problem we are having.

See e.g. for other cases:

https://github.com/cnaude/PurpleIRC/issues/141

https://github.com/StackStorm-Exchange/stackstorm-irc/issues/1

We need the bot to be authenticated as its registered nick to be able to join our channels. But as Freenode doesn't allow it to even connect, it is isn't possible to even get it to lurk and receive private messages.

Freenode SASL docs: http://freenode.net/kb/answer/registration http://freenode.net/kb/answer/sasl

Unfortunately Twisted IRC doesn't have the ability to use SASL auth.

### PR DESCRIPTION
**Pull request by ramiro**
https://github.com/gawel/irc3

It:

* Supports SASL auth
* Doesn't have other heavy baggage (like supybot/limnoria or even
  Twisted
* The bot can be created/configured programmatically (useful for Heroku
  config vars)
* Is maintained, e.g. is Python3-only
* Sane license

### CODE REVIEW

File: ticketbot.py [hunk 1]
```diff
@@ -73,85 +71,73 @@ def get_links(match_set, sha_validation=validate_sha_github):
...
-                "7 characters), subversion changesets like \"r12345\", and "
-                "github pull requests like \"PR12345\" or \"!12345\"."
+                "7 characters), and github pull requests like \"PR12345\" or \"!12345\"."
             )
-            self.msg(
+            self.bot.privmsg(
                 user,
                 "Suggestions? Problems? Help make me better: "
-                "https://github.com/django/django-ticketbot/"
+                "https://github.com/django/ticketbot/"
```
Comment (adamchainz) [inline ticketbot.py]: tiny thing: canonical github repo urls don't have a trailing slash

File: ticketbot.py [hunk 2]
```diff
@@ -73,85 +71,73 @@ def get_links(match_set, sha_validation=validate_sha_github):
...
 
         # No content? Send helptext.
-        if msg.startswith(self.nickname) and not any(matches):
-            self.msg(
+        if not any(matches) and (is_privmsg or data.startswith(self.bot.nick)):
+            self.bot.privmsg(
                 user,
                 "Hi, I'm Django's ticketbot. I know how to linkify tickets "
                 "like \"#12345\", github changesets like \"a00cf3d\" (minimum "
-                "7 characters), subversion changesets like \"r12345\", and "
```
Comment (adamchainz) [inline ticketbot.py]: well spotted that svn changesets aren't supported

File: requirements.txt [hunk 3]
```diff
@@ -1,2 +1,8 @@
-Twisted==17.9
+certifi==2017.11.5
```
Comment (timgraham) [inline requirements.txt]: Personally, I wouldn't pin the dependencies of the libraries that we use as I think that just adds more work to keep current.

File: ticketbot.py [hunk 4]
```diff
@@ -73,85 +71,73 @@ def get_links(match_set, sha_validation=validate_sha_github):
...
+        self.bot = bot
 
-    def privmsg(self, user, channel, msg):
-        """This will get called when the bot receives a message."""
-        user = user.split('!', 1)[0]
-        matches = get_matches(msg)
+    @irc3.event(irc3.rfc.PRIVMSG)
+    def process_msg_or_privmsg(self, mask, event, target, data, **kw):
+        """Detect special markers, reply with their respective links."""
+        # We shouldn't send automatic replies to notices
```
Comment (timgraham) [inline ticketbot.py]: Avoid use of "we" in comments per our [style guide](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style), e.g. “Don't send..."

File: ticketbot.py [hunk 5]
```diff
@@ -73,85 +71,73 @@ def get_links(match_set, sha_validation=validate_sha_github):
...
+        sasl_username=NICK,
+        sasl_password=password,
+        url='https://github.com/django/ticketbot',
+        autojoins=channels,
+        host='chat.freenode.net', port=6667, ssl=False,
+        includes=[
+            'irc3.plugins.core',
+            'irc3.plugins.sasl',
+            __name__,  # this register our Plugin
+            ],
```
Comment (timgraham) [inline ticketbot.py]: deindent

File: ticketbot.py [hunk 6]
```diff
@@ -73,85 +71,73 @@ def get_links(match_set, sha_validation=validate_sha_github):
...
-        connector.connect()
 
-    def clientConnectionFailed(self, connector, reason):
-        print("connection failed:", reason)
-        reactor.stop()
+def main():
+    password = os.environ['NICKSERV_PASS']
+    channels = os.environ['CHANNELS'].split(',')
+    # instantiate a bot
+    config = dict(
```
Comment (timgraham) [inline ticketbot.py]: omit the intermediate `config` variable?

File: ticketbot.py [hunk 7]
```diff
@@ -23,7 +21,7 @@
 
 
 MatchSet = namedtuple('MatchSet',
-    ['tickets', 'github_changesets', 'github_PRs'])
+                      ['tickets', 'github_changesets', 'github_PRs'])
```
Comment (timgraham) [inline ticketbot.py]: Following Django's style, we would prefer a single longer line (and we would also prefer not to make unrelated cosmetic changes in a PR that changes functionality, but whatever).

File: README.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+It needs the following env vars:
+
+* `NICKSERV_PASS` -- The bot's Freenode user password
+* `CHANNELS` -- A comma-separated list of channels it will auto-join to
+
+Example, for running it locally:
+
+```
+$ export NICKSERV_PASS=asecret
```
Comment (timgraham) [inline README.md]: asecret -> password?

Add a discussion of creating a virtualenv with the Python version specified in runtime.txt and installing requirements.txt?

File: ticketbot.py [hunk 9]
```diff
@@ -73,85 +71,73 @@ def get_links(match_set, sha_validation=validate_sha_github):
...
-        """If we get disconnected, reconnect to server."""
-        connector.connect()
 
-    def clientConnectionFailed(self, connector, reason):
-        print("connection failed:", reason)
-        reactor.stop()
+def main():
+    password = os.environ['NICKSERV_PASS']
+    channels = os.environ['CHANNELS'].split(',')
+    # instantiate a bot
```
Comment (timgraham) [inline ticketbot.py]: I think the comment isn't needed.

File: ticketbot.py [hunk 10]
```diff
@@ -73,85 +71,73 @@ def get_links(match_set, sha_validation=validate_sha_github):
...
+    def __init__(self, bot):
+        self.bot = bot
 
-    def privmsg(self, user, channel, msg):
-        """This will get called when the bot receives a message."""
-        user = user.split('!', 1)[0]
-        matches = get_matches(msg)
+    @irc3.event(irc3.rfc.PRIVMSG)
+    def process_msg_or_privmsg(self, mask, event, target, data, **kw):
+        """Detect special markers, reply with their respective links."""
```
Comment (timgraham) [inline ticketbot.py]: , -> and

File: README.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+$ export CHANNELS=#django-social,#django,#django-dev,#django-sprint,#django-core
+$ python ticketbot.py
+```
+
+## Tests
+
+There are some test cases for the text matching code. You can run them with
```
Comment (timgraham) [inline README.md]: with:

File: requirements.txt (hunk 3 above)
Comment (adamchainz) [reply requirements.txt]: but repeatable builds!

**Comment by ramiro** [comment]
Merged in 6395757d9a4db83eccaf54dbf92d3d2d951931b9.
